tests/integration/integration_test.go

```diff
@@ -800,6 +801,37 @@ func testConstructMethodsUnknown(t *testing.T, lang string, dependencies ...stri
 	}
 }
 
+func buildTestProvider(t *testing.T, providerDir string) string {
+	fn := func() {
+		providerName := "pulumi-resource-testprovider"
+		if runtime.GOOS == "windows" {
+			providerName += ".exe"
+		}
+
+		_, err := os.Stat(filepath.Join(providerDir, providerName))
+		if err == nil {
+			return
+		} else if errors.Is(err, os.ErrNotExist) {
+			// Not built yet, continue.
+		} else {
+			t.Fatalf("Unexpected error building test provider: %v", err)
+		}
+
+		cmd := exec.Command("go", "build", "-o", providerName)
+		cmd.Dir = providerDir
+		output, err := cmd.CombinedOutput()
+		if err != nil {
+			contract.AssertNoErrorf(err, "failed to run setup script: %v", string(output))
+		}
+	}
+	lockfile := filepath.Join(providerDir, ".lock")
```

Comment:
There's some options beside lockfiles, maybe, there func TestMain() like in https://medium.com/goingogo/why-use-testmain-for-testing-in-go-dafb52b406bc that has suite-level setup/teardown. IDK maybe lockfiles work fine, but this method would avoid tests contending on them.

Reply:
The lockfile ensures that package parallelism can be used.

Reply:
Yeah, `go test -p 2` is package parallelism, runs two test binaries side by side.

Found that worked really well to improve build times and Windows jobs were able to succeed with the reduced workload in smoke tests.